File: ResourceManager/Services/Abstractions/IDataService.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq.Expressions
Imports System.Threading.Tasks

Namespace ResourceManager.Services.Abstractions
    Public Interface IDataService(Of T)
        Function GetAll() As Task(Of IList(Of T))
        Function [Get](ByVal id As Integer) As Task(Of T)
        Function Create(ByVal entity As T) As Task(Of T)
        Function Update(ByVal id As Integer, ByVal entity As T) As Task(Of T)
        Function Delete(ByVal id As Integer) As Task(Of Boolean)
        Function GetWithRawSql(ByVal query As String, ParamArray parameters As Object()) As Task(Of IList(Of T))
        Function GetWhere(ByVal predicate As Expression(Of Func(Of T, Boolean))) As Task(Of IList(Of T))
        Function FirstOrDefault(ByVal predicate As Expression(Of Func(Of T, Boolean))) As Task(Of T)
        Function CountAll() As Task(Of Integer)
        Function CountWhere(ByVal predicate As Expression(Of Func(Of T, Boolean))) As Task(Of Integer)
        Function SingleOrDefault(ByVal predicate As Expression(Of Func(Of T, Boolean))) As Task(Of T)
    End Interface
End Namespace

File: ResourceManager/Services/Abstractions/IDataService.cs
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Threading.Tasks;

namespace ResourceManager.Services.Abstractions
{
    public interface IDataService<T>
    {
        Task<IList<T>> GetAll();
        Task<T> Get(int id);
        Task<T> Create(T entity);
        Task<T> Update(int id, T entity);
        Task<bool> Delete(int id);
        Task<IList<T>> GetWithRawSql(string query, params object[] parameters);
        Task<IList<T>> GetWhere(Expression<Func<T, bool>> predicate);
        Task<T?> FirstOrDefault(Expression<Func<T, bool>> predicate);
        Task<int> CountAll();
        Task<int> CountWhere(Expression<Func<T, bool>> predicate);
        Task<T?> SingleOrDefault(Expression<Func<T, bool>> predicate);
    }
}
